### ISSUE TITLE
Mockito.when() fails when method could originate from superclass or interface 

### ISSUE THREAD

**Issue by hansjoachim** [mockito/mockito#508]
I was testing/rebuilding a project with the latest mockito 2.0-beta when I noticed a couple of tests had started failing. Looked into why and was able to reduce the issue to [a reproducible test case]( https://github.com/hansjoachim/mockito/blob/confusedSignature/src/test/java/org/mockito/ConfusedSignatureTest.java). 

It looks like the problem is somehow caused by the method signature for `getFoo()`. The subclass gets this method from two places, the superclass with generic return type and from the interface with a concrete one, and this seems to be confusing Mockito. 

The test works with 1.10.19, but fails with 2.0.90-beta (/latest HEAD).

PS. Feel free to replace the title with something more descriptive.

**Comment by TimvdLippe** [comment]
I think this is an issue with Bytebuddy, WDYT @raphw?

**Comment by bric3** [comment]
@hansjoachim Have you tried other mockito 2 beta versions ?

**Comment by raphw** [comment]
Byte Buddy overrides the method as follows:

```java
public String getFoo() {
  return (String) MockMethodInterceptor$DispatcherDefaultingToRealMethod.interceptSuperCallable(
    mockitoInterceptor, 
    cachedMethod, 
    new Object[0],
    new Foo$Sub$MockitoMock$58829931$auxiliary$Ey71ZQFr(this)
  );
}
```

This is the expected value. Does not look like a Byte Buddy related problem to me.

**Comment by bric3** [comment]
After a `git bisect` the problem appeared with this commit 7fabd84fdfb2f80f8c135fe3865263c684c0d4e7


```
~/opensource/mockito ± ❯ git bisect bad
~/opensource/mockito ± ❯ git bisect good v2.0.26-beta
git bisect run ./gradlew :test --tests "org.mockitousage.bugs.ConfusedSignatureTest"

...

7fabd84fdfb2f80f8c135fe3865263c684c0d4e7 is the first bad commit
commit 7fabd84fdfb2f80f8c135fe3865263c684c0d4e7
Author: Rafael Winterhalter <rafael.wth@web.de>
Date:   Thu Nov 12 11:38:07 2015 +0100

    Updated to Byte Buddy 0.7: Adds support for preserving generic type information at runtime.

    For users of Mockito, this version brings the following advantages:
    1. A mocked method contains its generic type: A mocked method of type List<String> will not be erased to List. Many frameworks / libraries extract this information and would then react differently when encountering an instance of the mock class compared to the original class.
    2. Bridge methods can only be resolved appropriately when the generic type information is processed. This allows for a reliable recovery of "visibility bridges" which are no true bridges and are now reliably mocked.

:100644 100644 d27873a17388f63c683339dfaa640d35f0ac27ed 72f5bb3fbdf07ea04900c195c46ab3990f91135b M	build.gradle
:040000 040000 20affcb117604ba6251e1195c1820ffb09664cec aa31f07848d605ab4564587afa2c5b63daf5fca7 M	src
bisect run success
```

**Comment by raphw** [comment]
Strangely enough, when setting a breakpoint, the interceptor method does not seem to be invoked.

**Comment by bric3** [comment]
Yes it is weird. The issue seems to have been introduced by ByteBuddy 0.7

**Comment by bric3** [comment]
Things that matter, I'm using java 1.7

```
~/opensource/mockito ± ❯ java -version
java version "1.7.0_80"
Java(TM) SE Runtime Environment (build 1.7.0_80-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.80-b11, mixed mode)
```

**Comment by bric3** [comment]
@raphw Yup the test fails with `MissingMethodInvocationException`

**Comment by raphw** [comment]
I found the problem. The bridge method is no longer identified for some reason, it seems like Byte Buddy struggles to determine the bridge method's target.

**Comment by raphw** [comment]
I fixed it. Still cleaning up for a new Byte Buddy release. Should be fixed by tomorrow or latest Friday.

**Comment by bric3** [comment]
Thanks @raphw !

**Comment by hansjoachim** [comment]
Thank you, I've tested again and it's working now :)

I didn't see any tests added alongside the fix, and since it seems this bug was introduced a while back, I wonder if there should be one to catch regressions. If you'd like to, I can submit the test I made above in a PR. I wasn't really sure where to place the test, so if you have any suggestions for this or general style guidelines, just let me know.
